eudemo/eudemo/maintenance/equatorial_port.py

```diff
@@ -0,0 +1,229 @@
+from typing import Dict, List, Type, Union
+
+import numpy as np
+
+from bluemira.base.builder import Builder, ComponentManager
+from bluemira.base.components import Component, PhysicalComponent
+from bluemira.base.designer import Designer
+from bluemira.base.parameter_frame import Parameter, ParameterFrame
+from bluemira.builders.tools import circular_pattern_component
+from bluemira.display.palettes import BLUE_PALETTE
+from bluemira.geometry.plane import BluemiraFace, BluemiraPlane
+from bluemira.geometry.solid import BluemiraSolid, BluemiraWire
+from bluemira.geometry.tools import (
+    boolean_fuse,
+    extrude_shape,
+    make_polygon,
+    offset_wire,
+    slice_shape,
+)
+
+
+class EquatorialPort(ComponentManager):
+    """
+    Wrapper around a Equatorial Port component tree
+    """
+
+    def xz_boundary(self) -> BluemiraWire:
+        """Returns a wire defining the x-z boundary of the Equatorial Port"""
+        return (
+            self.component.get_component("xz")
+            .get_component(CastellationBuilder.NAME)
+            .shape.boundary[0]
+        )
+
+
+@dataclass
+class EquatorialPortDesignerParams(ParameterFrame):
+    """
+    Equatorial Port Designer parameters
+    """
+
+    ep_height: Parameter[float]
+
+
+@dataclass
+class CastellationBuilderParams(ParameterFrame):
+    """
+    Castellation Builder parameters
+    """
+
+    n_components: Parameter[int]
+    cst_r_corner: Parameter[float]
+
+
+class EquatorialPortDesigner(Designer):
+    """
+    Equatorial Port Designer
+    """
+
+    param_cls: Type[EquatorialPortDesignerParams] = EquatorialPortDesignerParams
+
+    def __init__(
+        self,
+        params: Union[Dict, ParameterFrame, EquatorialPortDesignerParams],
+        build_config: Union[Dict, None],
+        koz_z_offset: float,
+        x_ib: float,
+        x_ob: float,
+    ):
+        super().__init__(params, build_config)
+        self.koz_offset = koz_z_offset
+        self.x_ib = x_ib
+        self.x_ob = x_ob
+
+    def run(self) -> BluemiraWire:
+        """
+        Design the xz keep-out zone profile of the equatorial port
+        """
+        x_ib = self.x_ib
+        x_ob = self.x_ob
+        z_h = self.params.ep_height.value + (2 * self.koz_offset)
+
+        x = (x_ib, x_ob, x_ob, x_ib)
+        z = (-z_h / 2, -z_h / 2, z_h / 2, z_h / 2)
+
+        ep_boundary = BluemiraFace(
+            make_polygon({"x": x, "y": 0, "z": z}, label="koz", closed=True)
+        )
+        return ep_boundary
+
+
+class CastellationBuilder(Builder):
+    """
+    Castellation Builder
+    """
+
+    NAME = "Castellation"
+    param_cls: Type[CastellationBuilderParams] = CastellationBuilderParams
+
+    def __init__(
+        self,
+        params: Union[Dict, ParameterFrame, CastellationBuilderParams],
+        build_config: Union[Dict, None],
+        length: float,
+        start_profile: BluemiraFace,
+        extrude_vec: tuple,
+        offsets: list,
+        castellation_offsets: list,
```

Comment:
There's a fence-post problem in the suggestion, but that's easily solved by instead checking `sum(depth_offsets) < depth` so the final castellation isn't just a flared face at the port end. 

But this is somewhat solved by making the `depth_offsets` to be in absolute terms rather than relative. Then it's easy to check `depth_offsets[-1] < depth` and may make it clearer to the user when they've entered a final castellation depth that is at or beyond the total depth. I prefer absolute depths, especially since I imagine it could get tedious to adjust values if they're chained. The offset values I'm happy to keep relative since that's more intuitive and to reduce the risk a user designs a impractical/nonsense port plug where it castellates inwards and outwards.